spring-vaadin-security/src/main/java/org/vaadin/spring/security/provider/AbstractAnnotationAccessDelegate.java

```diff
@@ -0,0 +1,166 @@
+            return true;
+        } else if ( security.hasAccessDecisionManager() ) {
+
+        	logger.trace("DecisionManager present");
+        	
+            final Class<?> targetClass = AopUtils.getTargetClass(applicationContext.getBean(beanName));
+            final Method method = ClassUtils.getMethod(AopUtils.getTargetClass(applicationContext.getBean(beanName)), "enter", com.vaadin.navigator.ViewChangeListener.ViewChangeEvent.class);
+            final MethodInvocation methodInvocation = MethodInvocationUtils.createFromClass(targetClass, method.getName());
+
+            final Authentication authentication = security.getAuthentication();
+            final AccessDecisionManager accessDecisionManager = security.getAccessDecisionManager();
+            final ExpressionBasedAnnotationAttributeFactory attributeFactory = new ExpressionBasedAnnotationAttributeFactory(new DefaultMethodSecurityExpressionHandler());
+
+            Collection<ConfigAttribute> attributes = new ArrayList<ConfigAttribute>();
+            
+            try {
+            	
+            	logger.trace("Fetching value() from annotation");
+            	Object annotationValue = getAnnotationValue(targetClass, annotation, annotationAttribute);
+            	if ( annotationValue instanceof String[] ) {
+            		for(String securityValue : (String[]) annotationValue) {
+            			attributes.add(new SecurityConfig(securityValue));
+            		}
+            	} else {
+            		attributes.add(attributeFactory.createPreInvocationAttribute(null, null, annotationValue.toString()));
+            	}
+
+            	logger.trace("Requesting decision from decisionmanager");
+                accessDecisionManager.decide(authentication, methodInvocation, attributes);
+                return true;
+            } catch (InsufficientAuthenticationException e) {
+                return false;
+            } catch (AccessDeniedException e) {
+                return false;
+            } catch (Exception e) {
+				e.printStackTrace();
+				return false;
+			}
+
+        } else {
+        	logger.trace("Decision manager is required for @PreAuthorize; defaulting to 'true'");
+            return true; // Access decision manager required for @PreAuthorize()
+        }
+		
+	}
+	
+	@SuppressWarnings("unchecked")
+	protected static<T> T getAnnotationValue(Class<?> clazz, Class<? extends Annotation> annotationClass, String element) throws Exception {
+		Annotation annotation = clazz.getAnnotation(annotationClass);
+		Method method = annotationClass.getMethod(element, (Class[]) null);
+		if ( annotation == null ) {
+			return (T) method.getDefaultValue();
+		} else {
+			return (T) method.invoke(annotation, (Object[]) null);
+		}
+	}
+	
+	/**
+	 * If there is an access manager then the decision is already made
+	 */
+	@Override
+	public boolean isAccessGranted(View view, UI ui) {
+		return true;
```

Comment:
I'd actually prefer doing the security check that require the view instance here, because then it will work with any scope (including prototype scope). The view scope was merged into the official add-on today and I've suggested that this isAccessGranted(..) method should also get the bean name, in addition to the View and U instances. That way, you can use the bean name to ask the application context for any annotations you might need.